xunit.runner.wpf/ViewModel/MainViewModel.cs

```diff
@@ -204,135 +208,43 @@ private CommandBindingCollection CreateCommandBindings()
-            private readonly Func<bool> isCancelRequested;
-            private readonly IEnumerable<TestCaseViewModel> testCases;
-
-            public event EventHandler<TestStateEventArgs> TestFinished;
+                    session.TestDiscovered += OnTestDiscovered;
+                    session.SessionFinished += delegate { OnTestSessionFinished(session); };
 
-            public TestRunVisitor(IEnumerable<TestCaseViewModel> testCases, Func<bool> isCancelRequested)
-            {
-                this.testCases = testCases;
-                this.isCancelRequested = isCancelRequested;
-            }
-
-            protected override bool Visit(ITestFailed testFailed)
-            {
-                var testCase = testCases.Single(tc => tc.DisplayName == testFailed.TestCase.DisplayName);
-                testCase.State = TestState.Failed;
-                var resultString = new StringBuilder(testFailed.TestCase.DisplayName);
-                resultString.AppendLine(" FAILED:");
-                for (int i = 0; i < testFailed.ExceptionTypes.Length; i++)
-                {
-                    resultString.AppendLine($"\tException type: '{testFailed.ExceptionTypes[i]}', number: '{i}', parent: '{testFailed.ExceptionParentIndices[i]}'");
-                    resultString.AppendLine($"\tException message:");
-                    resultString.AppendLine(testFailed.Messages[i]);
-                    resultString.AppendLine($"\tException stacktrace");
-                    resultString.AppendLine(testFailed.StackTraces[i]);
+                    Assemblies.Add(new TestAssemblyViewModel(assembly));
                 }
-                resultString.AppendLine();
 
-                TestFinished?.Invoke(this, TestStateEventArgs.Failed(resultString.ToString()));
-                return !isCancelRequested();
+                this.IsBusy = true;
+                this.RunCommand.RaiseCanExecuteChanged();
+                this.CancelCommand.RaiseCanExecuteChanged();
+                await Task.WhenAll(this.testSessionList.Select(x => x.Task));
             }
-
-            protected override bool Visit(ITestPassed testPassed)
+            catch (Exception ex)
             {
-                var testCase = testCases.Single(tc => tc.DisplayName == testPassed.TestCase.DisplayName);
-                testCase.State = TestState.Passed;
-                TestFinished?.Invoke(this, TestStateEventArgs.Passed);
-                return !isCancelRequested();
+                MessageBox.Show(Application.Current.MainWindow, ex.ToString());
             }
-
-            protected override bool Visit(ITestSkipped testSkipped)
+            finally
             {
-                var testCase = testCases.Single(tc => tc.DisplayName == testSkipped.TestCase.DisplayName);
-                testCase.State = TestState.Skipped;
-                TestFinished?.Invoke(this, TestStateEventArgs.Skipped);
-                return !isCancelRequested();
+                loadingDialog.Close();
```

Comment:
Hmm.  I think a `Debug.Assert(!IsBusy)` would make sense here + a reset of `IsBusy` in the `catch` block.  That should cover all of the cases.